protocols/gossipsub/src/types.rs

```diff
@@ -19,18 +19,60 @@
 // DEALINGS IN THE SOFTWARE.
 
 //! A collection of types using the Gossipsub system.
+use crate::metrics::Metrics;
 use crate::TopicHash;
+use async_channel::{Receiver, Sender};
+use futures::{stream::Peekable, Stream, StreamExt};
+use futures_timer::Delay;
 use libp2p_identity::PeerId;
 use libp2p_swarm::ConnectionId;
 use prometheus_client::encoding::EncodeLabelValue;
 use quick_protobuf::MessageWrite;
 use std::fmt::Debug;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+use std::task::{Context, Poll};
+use std::time::Duration;
 use std::{collections::BTreeSet, fmt};
 
 use crate::rpc_proto::proto;
 #[cfg(feature = "serde")]
 use serde::{Deserialize, Serialize};
 
+/// The type of messages that have expired while attempting to send to a peer.
+#[derive(Clone, Debug, Default)]
+pub struct FailedMessages {
+    /// The number of publish messages that failed to be published in a heartbeat.
+    pub publish: usize,
+    /// The number of forward messages that failed to be published in a heartbeat.
+    pub forward: usize,
+    /// The number of messages that were failed to be sent to the priority queue as it was full.
+    pub priority: usize,
+    /// The number of messages that were failed to be sent to the non-priority queue as it was full.
+    pub non_priority: usize,
+}
+
+impl FailedMessages {
+    /// The total number of messages that expired due a timeout.
+    pub fn total_timeout(&self) -> usize {
+        self.publish + self.forward
+    }
+
+    /// The total number of messages that failed due to the queue being full.
+    pub fn total_queue_full(&self) -> usize {
+        self.priority + self.non_priority
+    }
+
+    /// The total failed messages in a heartbeat.
+    pub fn total(&self) -> usize {
+        self.total_timeout() + self.total_queue_full()
```

Comment:
Isn't for example in case of a failed published message both `priority` **and** `publish` increased, and thus it would count twice into this sum?